tests/std/tests/LWG3545_pointer_traits_sfinae/test.compile.pass.cpp

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cstddef>
+#include <memory>
+#include <type_traits>
+
+using namespace std;
+
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__);
+
+template <class, class = void>
+constexpr bool has_memtype_element_type = false;
+
+template <class T>
+constexpr bool has_memtype_element_type<T, void_t<typename T::element_type>> = true;
+
+template <class, class = void>
+constexpr bool has_memtype_pointer = false;
+
+template <class T>
+constexpr bool has_memtype_pointer<T, void_t<typename T::pointer>> = true;
+
+template <class, class = void>
+constexpr bool has_memtype_difference_type = false;
+
+template <class T>
+constexpr bool has_memtype_difference_type<T, void_t<typename T::difference_type>> = true;
+
+STATIC_ASSERT(!has_memtype_element_type<pointer_traits<int>>);
+STATIC_ASSERT(!has_memtype_pointer<pointer_traits<int>>);
+STATIC_ASSERT(!has_memtype_difference_type<pointer_traits<int>>);
+
+struct LackElementType {
+    using pointer         = int;
+    using difference_type = int;
+    template <class>
+    using rebind = int;
+};
+
+STATIC_ASSERT(!has_memtype_element_type<pointer_traits<LackElementType>>);
+STATIC_ASSERT(!has_memtype_pointer<pointer_traits<LackElementType>>);
+STATIC_ASSERT(!has_memtype_difference_type<pointer_traits<LackElementType>>);
+
+struct OnlyElementType {
+    using element_type = void;
+};
+
+STATIC_ASSERT(has_memtype_element_type<pointer_traits<OnlyElementType>>);
+STATIC_ASSERT(has_memtype_pointer<pointer_traits<OnlyElementType>>);
+STATIC_ASSERT(has_memtype_difference_type<pointer_traits<OnlyElementType>>);
+
+STATIC_ASSERT(is_same_v<pointer_traits<OnlyElementType>::element_type, void>);
+STATIC_ASSERT(is_same_v<pointer_traits<OnlyElementType>::pointer, OnlyElementType>);
+STATIC_ASSERT(is_same_v<pointer_traits<OnlyElementType>::difference_type, ptrdiff_t>);
+
+template <class I>
+struct Templated {
+    using difference_type = I;
+};
+
+STATIC_ASSERT(has_memtype_element_type<pointer_traits<Templated<char>>>);
+STATIC_ASSERT(has_memtype_pointer<pointer_traits<Templated<char>>>);
+STATIC_ASSERT(has_memtype_difference_type<pointer_traits<Templated<char>>>);
+
+STATIC_ASSERT(is_same_v<pointer_traits<Templated<char>>::element_type, char>);
+STATIC_ASSERT(is_same_v<pointer_traits<Templated<char>>::pointer, Templated<char>>);
+STATIC_ASSERT(is_same_v<pointer_traits<Templated<char>>::difference_type, char>);

```

Comment:
No change requested: Because this test case says `using difference_type = I;`, it's unable to distinguish whether `pointer_traits` is reading the template argument (as it did for `element_type`) or the `difference_type` typedef (as desired, which it is indeed doing). Using a unique type for the `difference_type`, like `int`, would make the test slightly stronger and clearer.

Below, `CheckPriority`'s `using element_type = T[42];` is a good example of detecting what `pointer_traits` is looking at.

I mention this only for future reference, as there is essentially zero risk in this specific case. This is related to my general recommendation for test data, which is "unique numbers for unique purposes".